File: app/schemas/rating.py
"""Pydantic schemas for rating and review system."""

from datetime import datetime
from uuid import UUID

from pydantic import Field, field_validator

from app.schemas.base import BaseSchema


class RatingCreate(BaseSchema):
    """Schema for creating a new rating."""

    booking_id: UUID
    overall_rating: int = Field(ge=1, le=5, description="Overall rating from 1-5 stars")

    # Optional category ratings
    professionalism_rating: int | None = Field(None, ge=1, le=5)
    punctuality_rating: int | None = Field(None, ge=1, le=5)
    care_of_items_rating: int | None = Field(None, ge=1, le=5)
    communication_rating: int | None = Field(None, ge=1, le=5)
    value_for_money_rating: int | None = Field(None, ge=1, le=5)

    # Review content
    review_text: str | None = Field(None, max_length=2000)
    review_title: str | None = Field(None, max_length=200)

    @field_validator("review_text")
    @classmethod
    def validate_review_text(cls, v: str | None) -> str | None:
        """Validate review text is not just whitespace."""
        if v and not v.strip():
            raise ValueError("Review text cannot be empty or just whitespace")
        return v.strip() if v else None


class RatingUpdate(BaseSchema):
    """Schema for mover response to rating."""

    mover_response: str = Field(max_length=1000, description="Mover's response to the review")

    @field_validator("mover_response")
    @classmethod
    def validate_response(cls, v: str) -> str:
        """Validate response is not empty."""
        if not v.strip():
            raise ValueError("Response cannot be empty")
        return v.strip()


class RatingResponse(BaseSchema):
    """Schema for rating response."""

    id: UUID
    booking_id: UUID
    org_id: UUID

    overall_rating: int
    professionalism_rating: int | None
    punctuality_rating: int | None
    care_of_items_rating: int | None
    communication_rating: int | None
    value_for_money_rating: int | None

    review_text: str | None
    review_title: str | None
    customer_name: str

    mover_response: str | None
    mover_responded_at: str | None

    is_published: bool
    is_verified_booking: bool

    created_at: datetime
    updated_at: datetime


class RatingSummaryResponse(BaseSchema):
    """Schema for rating summary statistics."""

    org_id: UUID
    total_ratings: int
    average_overall_rating: float

    # Category averages
    average_professionalism: float | None
    average_punctuality: float | None
    average_care_of_items: float | None
    average_communication: float | None
    average_value_for_money: float | None

    # Star distribution
    five_star_count: int
    four_star_count: int
    three_star_count: int
    two_star_count: int
    one_star_count: int

    created_at: datetime
    updated_at: datetime


class RatingListResponse(BaseSchema):
    """Schema for paginated rating list."""

    ratings: list[RatingResponse]
    total: int
    page: int
    page_size: int
    has_more: bool


class RatingStatsResponse(BaseSchema):
    """Schema for detailed rating statistics for an organization."""

    org_id: UUID
    summary: RatingSummaryResponse
    recent_ratings: list[RatingResponse]

    # Additional computed stats
    rating_trend: str  # "improving", "stable", "declining"
    response_rate: float  # Percentage of ratings with mover responses
    average_response_time_hours: float | None  # Average time to respond

File: frontend/src/types/rating.ts
/**
 * Rating and Review Types
 *
 * Type definitions for customer ratings and reviews
 */

export interface RatingCreate {
  booking_id: string;
  overall_rating: number; // 1-5 stars
  professionalism_rating?: number;
  punctuality_rating?: number;
  care_of_items_rating?: number;
  communication_rating?: number;
  value_for_money_rating?: number;
  review_text?: string;
  review_title?: string;
}

export interface Rating {
  id: string;
  booking_id: string;
  org_id: string;
  overall_rating: number;
  professionalism_rating: number | null;
  punctuality_rating: number | null;
  care_of_items_rating: number | null;
  communication_rating: number | null;
  value_for_money_rating: number | null;
  review_text: string | null;
  review_title: string | null;
  customer_name: string;
  mover_response: string | null;
  mover_responded_at: string | null;
  is_published: boolean;
  is_verified_booking: boolean;
  created_at: string;
  updated_at: string;
}

export interface RatingSummary {
  org_id: string;
  total_ratings: number;
  average_overall_rating: number;
  average_professionalism: number | null;
  average_punctuality: number | null;
  average_care_of_items: number | null;
  average_communication: number | null;
  average_value_for_money: number | null;
  five_star_count: number;
  four_star_count: number;
  three_star_count: number;
  two_star_count: number;
  one_star_count: number;
  created_at: string;
  updated_at: string;
}

export interface RatingListResponse {
  ratings: Rating[];
  total: number;
  page: number;
  page_size: number;
  has_more: boolean;
}

export interface RatingStats {
  org_id: string;
  summary: RatingSummary;
  recent_ratings: Rating[];
  rating_trend: 'improving' | 'stable' | 'declining';
  response_rate: number;
  average_response_time_hours: number | null;
}

export interface MoverResponse {
  mover_response: string;
}

export const RATING_CATEGORIES = [
  { key: 'professionalism', label: 'Professionalism' },
  { key: 'punctuality', label: 'Punctuality' },
  { key: 'care_of_items', label: 'Care of Items' },
  { key: 'communication', label: 'Communication' },
  { key: 'value_for_money', label: 'Value for Money' },
] as const;

export type RatingCategory = (typeof RATING_CATEGORIES)[number]['key'];
